Rocket.Chat/Controllers/Subscriptions/FilesListViewController.swift

```diff
@@ -0,0 +1,319 @@
+            completion?()
+        }
+    }
+}
+
+class FilesListViewController: BaseViewController {
+    var data = FilesListViewData()
+    var documentController: UIDocumentInteractionController?
+
+    @IBOutlet weak var tableView: UITableView!
+
+    @objc func refreshControlDidPull(_ sender: UIRefreshControl) {
+        let data = FilesListViewData()
+        data.subscription = self.data.subscription
+        data.loadMoreFiles {
+            self.data = data
+            DispatchQueue.main.async {
+                self.tableView.reloadData()
+                self.tableView.refreshControl?.endRefreshing()
+                self.updateIsEmptyFile()
+            }
+        }
+    }
+
+    func updateIsEmptyFile() {
+        guard let label = tableView.backgroundView as? UILabel else { return }
+
+        if data.cells.count == 0 {
+            label.text = localized("chat.messages.files.list.empty")
+        } else {
+            label.text = ""
+        }
+    }
+
+    func loadMoreFiles() {
+        data.loadMoreFiles {
+            DispatchQueue.main.async {
+                self.tableView.reloadData()
+                self.tableView.refreshControl?.endRefreshing()
+                self.updateIsEmptyFile()
+            }
+        }
+    }
+
+    func openImage(fromFile file: File, fromImageView imageView: FLAnimatedImageView) {
+        if imageView.animatedImage != nil || imageView.image != nil {
+            let configuration = ImageViewerConfiguration { config in
+                config.image = imageView.image
+                config.animatedImage = imageView.animatedImage
+                config.imageView = imageView
+                config.allowSharing = true
+            }
+            present(ImageViewerController(configuration: configuration), animated: true)
+        } else {
+            openRemoteImage(fromFile: file, fromImageView: imageView)
+        }
+    }
+
+    private func openRemoteImage(fromFile file: File, fromImageView imageView: FLAnimatedImageView?) {
+        guard let fileURL = file.fullFileURL() else { return }
+
+        func open(data: Data?) {
+            let configuration = ImageViewerConfiguration { config in
+                if let data = data {
+                    if file.isGif {
+                        config.animatedImage = FLAnimatedImage(gifData: data)
+                    } else {
+                        config.image = UIImage(data: data)
+                    }
+                }
+                config.imageView = imageView
+            }
+
+            DispatchQueue.main.async {
+                MBProgressHUD.hide(for: self.view, animated: true)
+                self.present(ImageViewerController(configuration: configuration), animated: false)
+            }
+        }
+
+        MBProgressHUD.showAdded(to: view, animated: true)
+        SDWebImageDownloader.shared().downloadImage(with: fileURL, options: [.useNSURLCache], progress: nil, completed: { _, data, _, _ in
```

Comment:
[weak self]?

Reply:
Honestly I don't know how it behave when we have a reference for `self` inside a sub function that is called inside an async-operation callback 🤔 I will update this sub function to a closure then capture the weak reference of self to make sure it will not cause any problem.